shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -293,6 +298,22 @@ std::vector<content::RenderFrameHost*> WebFrameMain::FramesInSubtree() const {
   return frame_hosts;
 }
 
+void WebFrameMain::Connect() {
+  if (pending_receiver_) {
+    render_frame_->GetRemoteInterfaces()->GetInterface(
+        std::move(pending_receiver_));
+  }
+}
+
+void WebFrameMain::OnDeleted() {
+  MarkRenderFrameDisposed();
+  Emit("destroyed");
```

Comment:
"destroyed" events have caused us some trouble in the past, especially when JS tries to do things related to the destroyed object. What's the use case for this event? Would it be reasonable to not run any JS here?

Reply:
I don't have any personal use case for it yet. I've added it to complement our frame-created event and since it seemed like it could be useful in the future.

I suppose if one needed to have frames subscribe to an event, you'd also want to know when they're removed so you can clear out their subscription.

I'd be interested to know whether @miniak had ideas as they also asked about including it.

Reply:
@nornagon how strongly do you feel about this? In the case of `WebFrameMain`, `CheckRenderFrame()` is called in each method/getter which throws upon access to a destroyed frame. As long as we maintain these checks, I think accessing `WebFrameMain` after being destroyed should be safe.

Reply:
@samuelmaddock you're probably right... the sort of thing i'm worried about is more complex though, e.g. what if `webContents.destroy()` is called from a `webFrameMain.on('destroyed')` callback? does that interfere with the destruction sequence in a bad way? Can that end up with things double-freed or UAFs or other nastiness?

Reply:
`RenderFrameHostImpl` contains a `RenderFrameState` field which can be set to `Deleting` and `Deleted`. It seems like there's some possibility of corruption if navigation occurs while the frame is in its `Deleting` state.
https://source.chromium.org/chromium/chromium/src/+/f9d298668873e378f1af2095039c011b9e43d732

`webFrameMain.on('destroyed')` is emitted while in this `Deleting` state. I can see two paths forward:
- Remove 'destroyed' and `isDestroyed()` until we have a solid use case for them.
- Defer `Emit('destroyed')` until the next tick? Do we have any precedence for this elsewhere?

Reply:
Let's go with the first option for now as the rest of this PR is still useful separately from a `destroyed` event.

Delaying event emission has other problems, we do occasionally do that sort of thing but I'd prefer to avoid it if we can. I wonder if we could somehow hook the _Deleted_ state instead of the _Deleting_...? But: let's consider that for a followup.

Reply:
@nornagon does it make sense to still include `WebFrameMain.isDestroyed()`? Then it could at least be tracked manually if needed.

Reply:
@samuelmaddock i'm on the fence about that one too, as I don't want to encourage polling to detect when the frame is destroyed. Let's pull out all the destroyed stuff into a separate PR.

Were you able to trigger that CHECK() btw?

Reply:
I wasn't able to trigger the CHECK(). I couldn't think of a way to trigger navigation or a reload after the `'destroyed'` event is emitted. `CheckRenderFrame()` prevents calling any such methods on the `WebFrameMain` instance.

Reply:
Removed the destroyed related code in 091b3f5

Reply:
Cool. Did you try navigating the parent WebContents by any chance?